src/driver/amdxdna/aie2_message.c

```diff
@@ -534,141 +536,216 @@ int aie2_execbuf(struct amdxdna_hwctx *hwctx, struct amdxdna_sched_job *job,
+		ret = aie2_cmdlist_fill_one_slot_cf(cmd_buf, offset, abo, size);
 		break;
 	case ERT_START_DPU:
-		msg.opcode = MSG_OP_CHAIN_EXEC_DPU;
-		ret = aie2_cmdlist_fill_slot_dpu(abo->mem.kva, job, &req.buf_size);
+		ret = aie2_cmdlist_fill_one_slot_dpu(cmd_buf, offset, abo, size);
 		break;
 	default:
 		ret = -EOPNOTSUPP;
 	}
+
+done:
 	if (ret) {
-		XDNA_ERR(xdna, "Failed to handle cmd op %d ret %d", op, ret);
-		return ret;
+		XDNA_ERR(abo->client->xdna, "Can't fill slot for cmd op %d ret %d",
+			 op, ret);
 	}
+	return ret;
+}
+
+static inline struct amdxdna_gem_obj *
+aie2_cmdlist_get_cmd_buf(struct amdxdna_sched_job *job)
+{
+	int idx = get_job_idx(job->seq);
 
-	XDNA_DBG(xdna, "Command buf addr 0x%llx size 0x%x count %d",
-		 req.buf_addr, req.buf_size, req.count);
+	return job->hwctx->priv->cmd_buf[idx];
+}
 
-	drm_clflush_virt_range(abo->mem.kva, req.buf_size);
-	/* Device can access the buf after flush */
+static inline void
+aie2_cmdlist_prepare_request(struct cmd_chain_req *req,
+			     struct amdxdna_gem_obj *cmdbuf_abo, u32 size, u32 cnt)
+{
+	req->buf_addr = cmdbuf_abo->mem.dev_addr;
+	req->buf_size = size;
+	req->count = cnt;
+	drm_clflush_virt_range(cmdbuf_abo->mem.kva, size);
+	XDNA_DBG(cmdbuf_abo->client->xdna, "Command buf addr 0x%llx size 0x%x count %d",
+		 req->buf_addr, size, cnt);
+}
 
-	msg.handle = handle;
+int aie2_cmdlist_multi_execbuf(struct amdxdna_hwctx *hwctx,
+			       struct amdxdna_sched_job *job,
+			       int (*notify_cb)(void *, const u32 *, size_t))
+{
+	struct amdxdna_gem_obj *cmdbuf_abo = aie2_cmdlist_get_cmd_buf(job);
+	struct mailbox_channel *chann = hwctx->priv->mbox_chann;
+	struct amdxdna_client *client = hwctx->client;
+	struct amdxdna_gem_obj *cmd_abo = job->cmd_bo;
+	struct amdxdna_cmd_chain *payload;
+	struct xdna_mailbox_msg msg;
+	struct cmd_chain_req req;
+	u32 payload_len;
+	u32 offset = 0;
+	u32 size;
+	int ret;
+	u32 op;
+	u32 i;
+
+	op = amdxdna_cmd_get_op(cmd_abo);
+	payload = amdxdna_cmd_get_payload(cmd_abo, &payload_len);
+	if (op != ERT_CMD_CHAIN || !payload ||
+	    payload_len < struct_size(payload, data, payload->command_count))
+		return -EINVAL;
+
+	for (i = 0; i < payload->command_count; i++) {
+		u32 boh = (u32)(payload->data[i]);
+		struct amdxdna_gem_obj *abo;
+
+		abo = amdxdna_gem_get_obj(client, boh, AMDXDNA_BO_CMD);
+		if (!abo) {
+			XDNA_ERR(client->xdna, "Failed to find cmd BO %d", boh);
+			return -ENOENT;
+		}
+
+		/* All sub-cmd should have same op, use the first one. */
+		if (i == 0)
+			op = amdxdna_cmd_get_op(abo);
+
+		ret = aie2_cmdlist_fill_one_slot(op, cmdbuf_abo, offset, abo, &size);
+		amdxdna_gem_put_obj(abo);
+		if (ret)
+			return -EINVAL;
+
+		offset += size;
+	}
+
+	aie2_cmdlist_prepare_request(&req, cmdbuf_abo, size, payload->command_count);
+
+	msg.handle = job;
 	msg.notify_cb = notify_cb;
 	msg.send_data = (u8 *)&req;
 	msg.send_size = sizeof(req);
+	ret = xdna_mailbox_send_msg(chann, &msg, TX_TIMEOUT);
```

Comment:
msg.opcode is not set??
Maybe
```suggestion
        if (op == ERT_START_CU)
             msg.opcode = MSG_OP_CHAIN_EXEC_BUFFER_CF;
        else
            msg.opcode = MSG_OP_CHAIN_EXEC_DPU;
	ret = xdna_mailbox_send_msg(chann, &msg, TX_TIMEOUT);
```

Reply:
Fixed.